### PR TITLE
Odd behaviour for S and P fixed in number of pi-bonding electrons calculation

### PR DESCRIPTION
**Pull request by kmnis**
`File modified: ` rdkit/rdkit/Chem/AtomPairs/Utils.py

Earlier version of the number of pi-bonded electrons calculation script behaved oddly for some of the molecules containing S and P. Below are two examples:

    **Old version**
    >>> m = Chem.MolFromSmiles('S(=O)(=O)(O)O')
    >>> NumPiElectrons(m.GetAtomWithIdx(0))
    0  # Wrong

    >>> m = Chem.MolFromSmiles('P(=O)(O)(O)O')
    >>> NumPiElectrons(m.GetAtomWithIdx(0))
    0  # Wrong

    **New version**
    >>> m = Chem.MolFromSmiles('S(=O)(=O)(O)O')
    >>> NumPiElectrons(m.GetAtomWithIdx(0))
    2

    >>> m = Chem.MolFromSmiles('P(=O)(O)(O)O')
    >>> NumPiElectrons(m.GetAtomWithIdx(0))
    1

This has been fixed in the modification. The formula used for the calculation is simply the difference of total bond order and number of non-hydrogen neighbors for non-aromatic atoms.

### CODE REVIEW

**Comment by greglandrum** [comment]
Thanks for the PR
The primary purpose of this `rdkit.Chem.AtomPairs.Utils.NumPiElectrons()` function is to return a value that is used in determining atom types for the atom pair and topological torsions fingerprints. As such, changing it would result in a change in the fingerprints themselves, which is a decision that needs to be made very carefully since it breaks backwards compatibility of results (and, in this case, also needs to be made in the C++ code, since that's what is now used to generate atom pair FPs).

In this case, I think it's probably not a good idea to change the fingerprints in order to make this particular correction.

Having a function to calculate the degree of unsaturation of an atom (which is more or less what you're doing here) *is* something that would be useful, perhaps in the `Chem.AllChem` module.

**Comment by sihagmnis36** [comment]
Yes, perhaps you're right. The problem I was working on had a high number of S and P atoms and hence I was getting lots of wrong results, so I thought of making the change. Should this function be moved to `AllChem.py` with the same or different function name, in case it's useful in some other context?

**Comment by greglandrum** [comment]
I think having a function in `AllChem.py` to calculate the degree of unsaturation of an atom could be quite useful. I'd be nervous about calling it `NumPiElectrons` since that can lead to arguments stemming from differences of opinion about what pi bonding actually means.
So maybe something like `AllChem.CalcCountAtomUnsaturations()`? That's not a great name, so maybe you have a better idea.

**Comment by sihagmnis36** [comment]
One doubt, when you said this is just calculation of the degree of unsaturation for an atom, I didn't get that. Isn't DoU is for molecules rather than individual atoms? Plus this is also not considering rings.

**Comment by greglandrum** [comment]
You're right that the usual definition is applied at the atomic level and it would probably be confusing to use it here (since we would not include the number of ring bonds).
We could call it CalcCountAtomUnsaturatedBonds()?

**Comment by sihagmnis36** [comment]
As we discussed, I have moved that piece of code to `AllChem.py` with the function name `GetNumOfUnsaturatedBonds` and restored the `NumPiElectrons` in `Utils.py`, back to original. Please take a look and check if everything's up to the mark.

**Comment by greglandrum** [comment]
The changes and tests look good. Thanks!

Looking through your doctests, it's clear to me that the name `GetNumOfUnsaturatedBonds()`, which is essentially what *I* suggested, is even more inaccurate than your original suggestion that it be `NumPiElectrons()`. For example, the Cs in `C#C` certainly only have one usaturated bond (for which they are using one pi electron), but the Cs in `c1ccccc1` each have two unsaturated bonds and a single pi electron. Sorry for having made the bad suggestion.

Let's go back to `GetNumPiElectrons()` and include a bit of additional text in the docstring. I will make a suggestion about this directly in the code.

File: rdkit/Chem/AllChem.py [hunk 1]
```diff
@@ -438,6 +439,65 @@ def AssignBondOrdersFromTemplate(refmol, mol):
...
+    1
+    
+    >>> GetNumOfUnsaturatedBonds(m.GetAtomWithIdx(3))
+    0
+    
+    >>> m = Chem.MolFromSmiles('c1ccccc1')
+    >>> GetNumOfUnsaturatedBonds(m.GetAtomWithIdx(0))
+    1
+    
+    >>> m = Chem.MolFromSmiles('S(=O)(=O)(O)O')
```
Comment (greglandrum) [inline rdkit/Chem/AllChem.py]: Perhaps add, just before these last two examples, something like:

For non-planar environments we return the number of multiple bonds that the atom is involved in (with triple bonds counting twice).

**Comment by sihagmnis36** [comment]
Yeah, it kinda all looks weird. Anyways, after giving it some thought, I think it isn't doing much but just returning the number of pi-electrons associated with an atom, taking care of both `S` and `P` atoms, although the method it uses seems unconventional and different from the original `GetNumPiElectrons` function. But this method can be verified by the original [Carhart et al.](https://pubs.acs.org/doi/10.1021/ci00046a002) paper on which atom pair feature calculations are based on. Since it's not available for the public, I'll just post the screenshot.

image

According to me, it'll be best suited to replace the original one in `Utils.py`. But your argument is also correct as it'll break the backward compatibility for Atom pair features. So perhaps the best solution that seems is to put it somewhere in comments and make replacement in future whenever there is revisiting of whole `rdkit C++` code. I've verified on multiple datasets that the function calculates pi-electrons accurately.

**Comment by greglandrum** [comment]
Closing due to lack of activity. Feel free to reopen if anyone has the time and inclination to revive this one.

**Pull request closed without merging**